### PR TITLE
consensus: change Quorum to use Certificates with RatificationResult

### PR DESCRIPTION
**Pull request by herr-seppia**
See also #1268

### CODE REVIEW

File: consensus/src/quorum/task.rs [hunk 1]
```diff
@@ -158,14 +158,22 @@ impl<'p, D: Database> Executor<'p, D> {
...
 
+            debug!(
+                event = "quorum_collected",
+                result = ?quorum.cert.result,
+                iter = quorum.header.iteration,
+                round = quorum.header.round,
+            );
+
             // Publish the quorum
             self.publish(msg.clone()).await;
```
Comment (goshawk-3) [inline consensus/src/quorum/task.rs:169]: Maybe we should (re)publish this only if `if let RatificationResult::Success(Vote::Valid(hash))` condition passes?

Comment (herr-seppia) [reply consensus/src/quorum/task.rs:169]: I believe we should broadcast any verified quorum.

Keep in mind that `RatificationResult::Fail` is still a valid quorum certificate, it means that ratification provisioners certified that the iteration must b skipped

Comment (fed-franz) [reply consensus/src/quorum/task.rs:169]: Seppia's right: we should propagate any valid quorum.

In this respect, do we add Failed certificates from quorum messages to our local Failed iterations list?

**Review by fed-franz — COMMENTED** [review]
Just a doubt to clarify on `verify_votes`

File: consensus/src/quorum/task.rs (hunk 1 above)
Comment (goshawk-3) [reply consensus/src/quorum/task.rs:169]: > Keep in mind that RatificationResult::Fail is still a valid quorum certificate, it means that ratification provisioners certified that the iteration must b skipped

Currently on collecting Quorum message, we process `RatificationResult::Success` case only, right? I don't see the logic of handling `RatificationResult::Fail" case implemented with this PR - no iteration is skipped?.

What if someone is flooding the network with the same `RatificationResult::Fail` Quorum and we constantly repropagate it? Is it a new vector attack?

Comment (herr-seppia) [reply consensus/src/quorum/task.rs:169]: We are validating that failed result some lines above

Comment (goshawk-3) [reply consensus/src/quorum/task.rs:169]: Lemme rephrase:
1.Where do you skip iterations due to valid quorum of type  `RatificationResult::Fail`? Where do you handle `RatificationResult::Fail`?

2. If an adversary floods network with valid quorum of type  `RatificationResult::Fail`, how does the node handle such attack?

Comment (herr-seppia) [reply consensus/src/quorum/task.rs:169]: Ok, get it

1. We cannot skip iteration from the quorum handler due to the current architecture (but I'm happy if you can correct me). The only thing we can do is to let the iteration timeout and include that `RatificationResult::Fail` in the failed_iterations
2. This doesn't seems an attack to me. Or at least is not a scenario introduced by this PR. Once that we solve the first point, this "attack" will become invalid

Comment (goshawk-3) [reply consensus/src/quorum/task.rs:169]: > We cannot skip iteration from the quorum handler due to the current architecture (but I'm happy if you can correct me). The only thing we can do is to let the iteration timeout and include that RatificationResult::Fail in the failed_iterations

Yes, you can skip an iteration. There are a few options to consider how to implement it. The easiest impl might be to have two handlers - one in Quorum::Task for `RatificationResult::Success` and one handler in MainTask::event_loop for `RatificationResult::Fail`. The dispatching will be in Chain component.

> This doesn't seems an attack to me. Or at least is not a scenario introduced by this PR. Once that we solve the first point, this "attack" will become invalid

By introducing `RatificationResult::Fail` Quorum we introduce a new vector attack. Anyway, if you believe there is no potential attack then it's fine.

Comment (fed-franz) [reply consensus/src/quorum/task.rs:169]: To clarify:
1. currently, quorum messages can only interrupt the whole round, so a Fail quorum is added to failed iterations but it does not stop the current iteration
2. Fail quorums should be handled like Success quorums: we only handle/propagate them if they match our current round/iteration

File: consensus/src/quorum/verifiers.rs [hunk 2]
```diff
@@ -135,13 +135,14 @@ pub fn verify_votes(
         target_quorum,
     };
 
-    if !quorum_result.quorum_reached() {
+    if vote != &Vote::NoQuorum && !quorum_result.quorum_reached() {
```
[the remark this answers is not in the document]
Comment (herr-seppia) [reply consensus/src/quorum/verifiers.rs:138]: you're right, updated

**Review by fed-franz — APPROVED** [review]

**Pull request merged**